### ISSUE TITLE
Use of Gauge

### ISSUE THREAD

**Issue by ihabadly** [Mikhus/canvas-gauges#56]
I have a problem when instantiating a repeated instances of the gauge, using something like that in the ID field is not working:
```<canvas canvas-gauge id="gauge-{{$index}}" .......></canvas>```

**Comment by Mikhus** [comment]
Not clear what are you going to achieve with your example and where you've found information about the way you wrote it. Please, refer [documentation](https://canvas-gauges.com/documentation/user-guide/using-as-component) for information about proper use.

**Comment by ihabadly** [comment]
I am using AngularJS wrapper from https://github.com/dec/angular-canvas-gauge, it just wraps all the parameters and make them accessible through AngularJS. In my markup I am trying to use ngRepeat to draw a number of gauges all together through:
    <div data-ng-repeat="gaugeitem in gauge">
        <canvas canvas-gauge id="gauge_{{$index}}" 
                width="{{gaugeitem.width}}" 
                height="{{gaugeitem.height}}" 
                data-value={{gaugeitem.value}} 
            ></canvas>
        </div>
However, when calling the new RadialGauge() and it finds such id name "gauge_{{$index}}" in the markup, it gives an error: "Error: Given gauge canvas element is invalid!"

**Comment by Mikhus** [comment]
This wrapper does not work with canvas-gauges v2.x.x. There is a corresponding [issue](https://github.com/dec/angular-canvas-gauge/issues/6).

**Comment by ihabadly** [comment]
I fixed all these issues and the wrapper now works with canvas-gauges v2.x.x in all scenarios except with the ngRepeat from AngularJS because of the unique ID that has to be given for every Gauge.

**Comment by Mikhus** [comment]
I'm not that familiar with angular wrapper, better to ask @dec (author) to help with that. From other hand if you have fixed issues with wrapper you can propose PR in corresponding repo. What I can assume here is that gauges tried to initialize earlier than angular did it's job, so at the initialization there is a couple of gauges with the same id ("gauge_{{$index}}") on a page and it's triggers error. I need to investigate this to give a precise answer.

**Comment by ihabadly** [comment]
I think the problem is related to document.getElementById it goes to the renderTo but since it is not string it throws that error.
Thanks for your efforts and time. I really appreciate your Gauge design.

**Comment by Mikhus** [comment]
With the latest release, after investigations I can tell that you no longer need to use specific wrapper to integrate gauges into angular 1 or 2 applications, as far as gauges themselves are fully-functional web components since version 2. There was several problems which causes them crash in some circumstances (for example, when calling on angular's digest cycles). But with version 2.0.4 all that problems are fixed.

So you may simply refer gauges code from your html or by bundling their code into your script bundle within your build system.

Than you may simply use it in html.

## Use with angular 1.x

For example, this worked for me in angular1.5 app well:

```javascript
'use strict';

angular.module('myApp.view1', ['ngRoute'])

.config(['$routeProvider', function($routeProvider) {
  $routeProvider.when('/view1', {
    templateUrl: 'view1/view1.html',
    controller: 'View1Ctrl'
  });
}])

.controller('View1Ctrl', ['$scope', function($scope) {
  $scope.gauges = [{
    width: 150,
    height: 150
  },{
    width: 250,
    height: 250
  },{
    width: 350,
    height: 350
  },{
    width: 450,
    height: 450
  }];
}]);
```

where view1.html is something like:

```html
<canvas ng-repeat="gauge in gauges track by $index" data-type="radial-gauge"
        data-width="{{gauge.width}}"
        data-height="{{gauge.height}}"
        data-min-value="0"
        data-max-value="360"
        data-major-ticks="N,NE,E,SE,S,SW,W,NW,N"
        data-minor-ticks="22"
        data-ticks-angle="360"
        data-start-angle="180"
        data-stroke-ticks="false"
        data-highlights="false"
        data-color-plate="#33a"
        data-color-major-ticks="#f5f5f5"
        data-color-minor-ticks="#ddd"
        data-color-numbers="#ccc"
        data-color-needle="rgba(240, 128, 128, 1)"
        data-color-needle-end="rgba(255, 160, 122, .9)"
        data-value-box="false"
        data-value-text-shadow="false"
        data-color-circle-inner="#fff"
        data-color-needle-circle-outer="#ccc"
        data-needle-circle-size="15"
        data-needle-circle-outer="false"
        data-animation-rule="linear"
        data-needle-type="line"
        data-needle-start="75"
        data-needle-end="99"
        data-needle-width="3"
        data-borders="true"
        data-border-inner-width="0"
        data-border-middle-width="0"
        data-border-outer-width="10"
        data-color-border-outer="#ccc"
        data-color-border-outer-end="#ccc"
        data-color-needle-shadow-down="#222"
        data-border-shadow-width="0"
        data-animation-target="plate"
        data-units="ᵍ"
        data-title="DIRECTION"
        data-font-title-size="19"
        data-color-title="#f5f5f5"
        data-animation-duration="1500"
        data-value="45"
        data-animate-on-init="true"
></canvas>
```

and gauges are simply included in the main html page, like:

```html
<script src="https://cdn.rawgit.com/Mikhus/canvas-gauges/gh-pages/download/latest/all/gauge.min.js"></script>
```

You may refer gauges code locally, of course or bundle them together with your code as I said before. Async loading is also allowed as well.

## Use with Angular 2

Just install gauges via npm:

    $ npm install canvas-gauges

And typings.  
 - For TypeScript < 2:
   ```typings install --save --source=dt --global canvas-gauges```
 - For TypeScript 2+
   ```npm install @types/canvas-gauges``` 

Than imagine, for example, such a simple `app.component.ts`

```typescript
import { Component } from '@angular/core';
import 'canvas-gauges';

@Component({
    selector: 'my-app',
    template: `
      <h1>My First Angular App</h1>
      <canvas data-type="radial-gauge"
              data-width="400"
              data-height="400"
      ></canvas>
    `
})
export class AppComponent { }
```

Hope, this helps.

**Comment by dec** [comment]
Good to know that @Mikhus! I have some problems when try to integrate the 2.0 version of the Gauge library into Angular 1.x, just as I do before with the 1.x branch of the Gauge library. But probably you have reason and we no need to add any wrapper above the Gauge library to be used into Angular. I want to try it too when possible. Thanks for your efforts @Mikhus!

**Comment by ihabadly** [comment]
Thank you very much @Mikhus , I tried it but it seems that the passing of data-width and data-height are not handled correctly. When I change them in the controller, they keep showing with other default values.

The data-major-ticks when used throws the following error: "TypeError: t.majorTicks.push is not a function"  
However other parameters as data-value and data-title are working just perfect.

**Comment by Mikhus** [comment]
@ihabadly, I'll check it, thank you for contribution

**Comment by Mikhus** [comment]
@ihabadly, I did a small angular1 example: [cgng1.zip](https://github.com/Mikhus/canvas-gauges/files/491820/cgng1.zip)
I can't see there any problem with updating width and height of the gauges via controller. So, seems the problem is elsewhere in your code or, please, provide an example where the problem can be reproduced.

To run my example it is enough to download, unzip and run

```
$ npm install
$ npm start
```

Then just open in browser `http://localhost:9000`

Thank you in advance.

**Comment by ihabadly** [comment]
Thank you very much for the effort and for the illustrative video. When I run application at my side it throws that error, may be it has something to with FireFox
![screenshot](https://cloud.githubusercontent.com/assets/22375309/18829137/3841b444-83db-11e6-8281-d5f2877d58ce.png)

**Comment by Mikhus** [comment]
@ihabadly , true story!
This has been fixed in release 2.0.5

You can check it with the same example, the only thing is you will need to change path to gauge code in index.html to: `//cdn.rawgit.com/Mikhus/canvas-gauges/gh-pages/download/2.0.5/all/gauge.min.js`

**Comment by ihabadly** [comment]
Thank you very much, you made me day.
Now I am only stuck with the majorTicks error:
![screenshot2](https://cloud.githubusercontent.com/assets/22375309/18847709/86e3b188-842b-11e6-916f-c394ee7a42f6.png)

**Comment by Mikhus** [comment]
@ihabadly, I really need some details on this. Which steps do I need to take to reproduce this. I don't face that issue in my example under Firefox, so, please, give me details.

**Comment by Mikhus** [comment]
Fixed in release 2.0.6
